evm_arithmetization/src/poseidon/poseidon_stark.rs

```diff
@@ -528,34 +989,135 @@ mod tests {
 
     #[test]
     fn poseidon_correctness_test() -> Result<()> {
+        let input: Vec<u8> = (0..POSEIDON_SPONGE_RATE * FELT_MAX_BYTES)
+            .map(|_| rand::random())
+            .collect();
+        let int_inputs = PoseidonOp::PoseidonGeneralOp(PoseidonGeneralOp {
+            base_address: MemoryAddress::new(
+                0,
+                crate::memory::segments::Segment::AccessedAddresses,
+                0,
+            ),
+            input: input.clone(),
+            timestamp: 0,
+            len: POSEIDON_SPONGE_RATE * FELT_MAX_BYTES,
+        });
         const D: usize = 2;
-        type C = PoseidonGoldilocksConfig;
-        type F = <C as GenericConfig<D>>::F;
+        type F = GoldilocksField;
         type S = PoseidonStark<F, D>;
 
         let stark = S {
             f: Default::default(),
         };
 
-        let input = PoseidonOp(F::rand_array());
-        let rows = stark.generate_trace_rows(vec![input], 8);
+        let rows = stark.generate_trace_rows(vec![int_inputs], 8);
         assert_eq!(rows.len(), 8);
-        let row: PoseidonColumnsView<F> = rows[0].into();
-        let expected = F::poseidon(input.0);
-        assert_eq!(
-            std::array::from_fn::<_, 4, _>(
-                |i| row.digest[2 * i] + row.digest[2 * i + 1] * F::from_canonical_u64(1 << 32)
-            ),
-            expected[0..POSEIDON_DIGEST]
-        );
+        let last_row: &PoseidonColumnsView<_> = rows[0].borrow();
+        let mut output: Vec<_> = (0..POSEIDON_DIGEST)
+            .map(|i| {
+                last_row.digest[2 * i] + F::from_canonical_u64(1 << 32) * last_row.digest[2 * i + 1]
+            })
+            .collect();
+
+        let hash = poseidon_hash_padded_byte_vec(input);
+
+        // TODO: Without the maps it says: "can't compare `GoldilocksField` with
+        // `plonky2_field::goldilocks_field::GoldilocksField`""
```

Comment:
This is because you're most likely having distinct versions of `plonky2` internally, the map shouldn't be necessary. Although I believe we perform a silent modular reduction when comparing elements. Here, we want exact values, so probably have to keep it as is, or call `x.to_noncannonical_u64()` instead of `x.0` eventually.

Reply:
As you say, the mismatch is because smt_trie uses `plonky2 = { git = "https://github.com/0xPolygonZero/plonky2", rev = "c1728d4e43e9ff434f9297e4f6171ddf28ec8fca" }` while evm_arithmethization uses `plonky2 = "0.2.1"`. I changed `smt_trie/Cargo.toml` to match that of evm_arithmetization, but not sure it would have any unforeseen consequences :boom:

Reply:
It shouldn't, probably a leftover from when i ported it from the separate repo